pkg/scheduler/schedule_one.go

```diff
@@ -1074,9 +1074,18 @@ func (sched *Scheduler) handleSchedulingFailure(ctx context.Context, fwk framewo
 			// ignore this err since apiserver doesn't properly validate affinity terms
 			// and we can't fix the validation for backwards compatibility.
 			podInfo.PodInfo, _ = framework.NewPodInfo(cachedPod.DeepCopy())
-			if err := sched.SchedulingQueue.AddUnschedulableIfNotPresent(logger, podInfo, sched.SchedulingQueue.SchedulingCycle()); err != nil {
-				logger.Error(err, "Error occurred")
-			}
+			schedulingCycle := sched.SchedulingQueue.SchedulingCycle()
+			go func() {
+				// When handling events takes time in the queue, AddUnschedulableIfNotPresent could be blocked by a shared lock within the queue,
+				// which negatively impacts the scheduling throughput.
+				// So, we put an unschedulable pod back to the queue in a goroutine.
+				//
+				// We handle a nominatingInfo outside goroutine below,
+				// and hence there shouldn't be any race condition etc that caused by doing this in goroutine.
+				if err := sched.SchedulingQueue.AddUnschedulableIfNotPresent(logger, podInfo, schedulingCycle); err != nil {
```

Comment:
> Better to add context here

You meant, to the parameter of `AddUnschedulableIfNotPresent`? 
If Yes, how do you want to use ctx over there?

> count the goroutine number metric.

This is good point.

Reply:
> count the goroutine number metric.

Done in the latest commit.

Reply:
> If Yes, how do you want to use ctx over there?

Just to make sure we always have a way to GC the goroutine in time, but seems we can't here because the scheduleCtx will be cancelled in the end. :(